compiler/codegen/symbol.cpp

```diff
@@ -1803,6 +1865,32 @@ GenRet FnSymbol::codegenPYXType() {
   return ret;
 }
 
+std::string FnSymbol::getPythonArrayReturnStmts() {
+  INT_ASSERT(retType == dtExternalArray);
+  Symbol* eltType = elementType[this];
+  std::string typeStr = getPythonTypeName(eltType->type, PYTHON_PYX);
+  std::string typeStrCDefs = typeStr;
+  if (strncmp(typeStr.c_str(), "numpy", strlen("numpy")) == 0) {
+    typeStrCDefs += "_t";
+  } else {
+    typeStrCDefs = getPythonTypeName(eltType->type, PYTHON_PXD);
+  }
+  // Create the numpy array to return
+  // E.g. cdef numpy.ndarray [C element type, ndim=1] ret =
+  //          numpy.zeros(shape = ret_arr.size, dtype = Python element type)
+  std::string res = "\tcdef numpy.ndarray [" + typeStrCDefs + ", ndim=1] ret";
+  res += " = numpy.zeros(shape = ret_arr.size, dtype = " + typeStr + ")\n";
+
+  // Populate it with the contents we return (which translated C types into
+  // Python types)
+  res += "\tfor i in range(ret_arr.size):\n";
+  res += "\t\tret[i] = (<" + typeStrCDefs + "*>ret_arr.elts)[i]\n";
```

Comment:
How many times do we copy these arrays? I thought one of the points of numpy arrays was that they could refer to arrays allocated in C. I think reducing the number of copies on this interface will be important for performance but obviously it can be put off.

Reply:
I tried a couple of things prior to this implementation that I had difficulty with.  I intend to return to them at some point, but figured slower support was better than no support.

- I tried experimenting with Cython's [memory views](http://docs.cython.org/en/latest/src/userguide/memoryviews.html) but was having trouble with getting it to play nice due to type stuff.

- I also tried initializing the numpy array's buffer to reference `elts` but hadn't fully wrapped my head around how to store function pointers from C into a Python type

Reply:
> I also tried initializing the numpy array's buffer to reference elts but hadn't fully wrapped my head around how to store function pointers from C into a Python type

Since we know how to call a deallocator, can we use a function pointer to the appropriate Python "free" function (which might decrement a reference count, after all it's Python)?

Anyway, probably worth an issue if you don't sort it out soon. I think there are other places with unnecessary copies.